Lib/test/test_tabnanny.py

```diff
@@ -0,0 +1,348 @@
+        """
+        file_path = create_tmp_python_file(SOURCE_CODES["nannynag_errored"])
+
+        out = "%r: *** Line 3: trouble in tab city! ***\n" % (file_path,)
+        out += "offending line: '\\tprint(\"world\")\\n'\n"
+        out += "indent not equal e.g. at tab size 1\n"
+
+        tabnanny.verbose = 1
+        self._verify_tabnanny_check(file_path, out=out)
+
+    def test_when_nannynag_error(self):
+        """A python source code file eligible for raising `tabnanny.NannyNag`.
+        """
+        file_path = create_tmp_python_file(SOURCE_CODES["nannynag_errored"])
+        out = "%s 3 '\\tprint(\"world\")\\n'\n" % (file_path,)
+        self._verify_tabnanny_check(file_path, out=out)
+
+    def test_when_no_file(self):
+        """A python file which does not exist actually in system."""
+        file_path = 'no_file.py'
+        err = "%r: I/O Error: [Errno 2] No such file or directory: %r\n" % (
+            file_path, file_path
+        )
+        self._verify_tabnanny_check(file_path, err=err)
+
+    def test_errored_directory(self):
+        """Directory containing wrongly indented python source code files."""
+        with TemporaryDirectory() as tmp_dir:
+            e_file = create_tmp_python_file(
+                SOURCE_CODES["wrong_indented"], dir=tmp_dir
+            )
+            c_file = create_tmp_python_file(SOURCE_CODES["error_free"], dir=tmp_dir)
+            err = ('unindent does not match any outer indentation level'
+                        ' (<tokenize>, line 3)\n')
+            err = "%r: Indentation Error: %s" % (e_file, err)
+            self._verify_tabnanny_check(tmp_dir, err=err)
+
+
+class TestProcessTokens(TestCase):
+    """Testing `tabnanny.process_tokens()`."""
+
+    @patch('tabnanny.NannyNag')
+    def test_with_correct_code(self, MockNannyNag):
+        """A python source code without any whitespace related problems."""
+
+        file_path = create_tmp_python_file(SOURCE_CODES["error_free"])
+        with open(file_path) as f:
+            tabnanny.process_tokens(tokenize.generate_tokens(f.readline))
+        self.assertFalse(MockNannyNag.called)
+
+    def test_with_errored_codes_samples(self):
+        """A python source code with whitespace related sampled problems."""
+
+        # "tab_space_errored_1": executes block under type == tokenize.INDENT
+        #                        at `tabnanny.process_tokens()`.
+        # "tab space_errored_2": executes block under
+        #                        `check_equal and type not in JUNK` condition at
+        #                        `tabnanny.process_tokens()`.
+
+        for key in ["tab_space_errored_1", "tab_space_errored_2"]:
+            file_path = create_tmp_python_file(SOURCE_CODES[key])
+            exception = tabnanny.NannyNag
+            with open(file_path) as f:
+                tokens = tokenize.generate_tokens(f.readline)
+                self.assertRaises(exception, tabnanny.process_tokens, tokens)
```

Comment:
Move this assert out of the with block.

Reply:
Why not? Is `tokenize.generate_tokens()` lazy?

Reply:
@brettcannon The `tokenize.generate_tokens()` is a lazy function (generator). Can you help to identify a way which allows putting `assertRaises` outside the `with`? Thanks!